zcash_primitives/src/zip32.rs

```diff
@@ -43,41 +48,148 @@ impl ConditionallySelectable for AccountId {
 
 // ZIP 32 structures
 
-/// A child index for a derived key
+/// A marker trait indicating whether a [`ChildIndex`] supports non-hardened derivation or
+/// not.
+pub trait Hardening: private::Sealed {}
+
+mod private {
+    pub trait Sealed: Sized {
+        fn master() -> Self;
+        fn hardened(index: u32) -> Self;
+        fn is_hardened(&self) -> bool;
+        fn index(&self) -> u32;
+    }
+}
+
+/// Marker indicating that a [`ChildIndex`] is always hardened.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Hard(u32);
+
+impl private::Sealed for Hard {
+    fn master() -> Self {
+        // Hardened-only derivation still uses the non-hardened child index of 0 for the
+        // master key, for historic reasons. This only appears in the extended spending
+        // key encoding format, and is not used anywhere.
+        Self(0)
+    }
+
+    fn hardened(index: u32) -> Self {
+        Self(index)
```

Comment:
This can only be implemented in this module, but it can be invoked from anywhere; shouldn't this method be fallible with a restricted range of possible inputs?

Reply:
This method was on a sealed trait. Users outside the crate can't name the trait, so they can't use the method.

The trait has now been removed.